src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs

```diff
@@ -656,6 +656,12 @@ public override bool HandleCall(MethodIL callingMethodBody, MethodDesc calledMet
                         if (Intrinsics.GetIntrinsicIdForMethod(callingMethodDefinition) == IntrinsicId.RuntimeReflectionExtensions_GetMethodInfo)
                             break;
 
+                        if (param.IsEmpty())
+                        {
+                            // The static value is unknown and the below `foreach` won't execute
```

Comment:
Why is the value unknown in my repro?

I would expect that the dataflow should be able to trace through `new Action(Test<string>).Method` without problems.

Reply:
Known limitation of the dataflow analysis (#93720 is the tracking issue).

Reply:
> I thought #93720 was about not tracking variable types for `GetType` - how does that explain this issue? If we don't trace through `new Action(Test<string>)`, I would have expected it to show up as an unknown value, not empty.

I thought unknown/empty are interchangeable but now I understand the difference. We eventually want to know the type allocated through the `new Action` do that `GetType` (on in this case Delegate.get_Method`) know the type it's operating on.

In this case, we end up with empty instead of unknown because constructors take this path (they "return void"):

https://github.com/dotnet/runtime/blob/3d0da2cc430333160de47b8345609c75830953ad/src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs#L1168-L1170

And we can no longer bash it to unknown here because the HandleCallAction returned true (i.e. `handledFunction`):

https://github.com/dotnet/runtime/blob/3d0da2cc430333160de47b8345609c75830953ad/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/MethodBodyScanner.cs#L1302-L1319

Reply:
The compiler keeps track of delegate type + method. If we know delegate type, we know all the methods it can be used with.

We don't track this at method granularity because doing `new Action(Foo).Method` is very rare. People only do this if they want to work around lack of `methodof` in C#. What happens 95% of time is that we construct a delegate somewhere and a different part of the program will call `Delegate.get_Method` on it. That's why we only care about delegate type.

Dataflow losing track of the type in the trivial `new Action(Foo).Method` pattern is throwing a wrench in this because if someone uses this to work around `methodof`, we only see `Delegate.get_Method` was called on some unknown thing and we need to disable the optimization globally because this could be any delegate.